node/forest_libp2p/src/service.rs

```diff
@@ -63,57 +83,76 @@ impl Libp2pService {
+    /// Starts the `Libp2pService` networking stack. This Future resolves when shutdown occurs.
+    pub async fn run(self) -> Result<(), ()> {
+        enum MergeEvent {
+            Swarm(MyBehaviourEvent),
+            Pubsub((Topic, Vec<u8>)),
+        };
+        let mut swarm_stream = self.swarm.map(MergeEvent::Swarm).fuse();
+        let mut pubsub_stream = self.pubsub_receiver_in.map(MergeEvent::Pubsub).fuse();
         loop {
-            match self.swarm.poll() {
-                Ok(Async::Ready(Some(event))) => match event {
+            select! {
+                swarm_event = swarm_stream.next() => match swarm_event {
+                    Some(MergeEvent::Swarm(event)) => match event {
                     MyBehaviourEvent::DiscoveredPeer(peer) => {
-                        libp2p::Swarm::dial(&mut self.swarm, peer);
+                        libp2p::Swarm::dial(&mut swarm_stream.get_mut().get_mut(), peer);
                     }
                     MyBehaviourEvent::ExpiredPeer(_) => {}
                     MyBehaviourEvent::GossipMessage {
                         source,
                         topics,
                         message,
                     } => {
-                        return Ok(Async::Ready(Option::from(NetworkEvent::PubsubMessage {
+                        info!(self.log, "Got a Gossip Message from {:?}", source);
+                        self.pubsub_sender_out.send(NetworkEvent::PubsubMessage {
                             source,
                             topics,
-                            message,
-                        })));
+                            message
+                        }).await;
                     }
+                    }
+                    Some(MergeEvent::Pubsub(_)) => error!(self.log, "Should not happen"),
+                    None => ()
                 },
-                Ok(Async::Ready(None)) => break,
-                Ok(Async::NotReady) => break,
-                _ => break,
-            }
+                pubsub_event = pubsub_stream.next() => match pubsub_event {
+                    Some(MergeEvent::Pubsub(payload)) => {
+                       swarm_stream.get_mut().get_mut().publish(&payload.0, payload.1);
+                    }
+                    Some(MergeEvent::Swarm(_)) => error!(self.log, "Should not happen"),
+                    None => ()
+                }
+            };
         }
-        Ok(Async::NotReady)
     }
-}
 
-/// Events emitted by this Service to be listened by the NetworkService.
-#[derive(Clone)]
-pub enum NetworkEvent {
-    PubsubMessage {
-        source: PeerId,
-        topics: Vec<TopicHash>,
-        message: Vec<u8>,
-    },
+    /// Returns a `Sender` allowing you to send messages over GossipSub
+    pub fn pubsub_sender(&self) -> Sender<(Topic, Vec<u8>)> {
+        self.pubsub_sender_in.clone()
+    }
+
+    /// Returns a `Receiver` to listen to GossipSub messages
+    pub fn pubsub_receiver(&self) -> Receiver<NetworkEvent> {
```

Comment:
is there a reason you don't pass a reference for these?

Reply:
You mean return a reference? These are multiple producer, multiple consumer channels. So we can clone these and pass them around

Reply:
Yeah that is what I meant, but yeah okay if cloning won't cause data inconsistencies and overhead. Just more wondering the use case for this because I don't see why you would need more than a reference to this (this function isn't used yet)